tlsutil/config.go

```diff
@@ -135,96 +130,106 @@ func SpecificDC(dc string, tlsWrap DCWrapper) Wrapper {
+// *tls.Config necessary for Consul. Except the one in the api package.
+type Configurator struct {
+	sync.RWMutex
+	base    *Config
+	cert    *tls.Certificate
+	cas     *x509.CertPool
+	logger  *log.Logger
+	version int
+}
 
-	// If crypto/tls is doing verification, there's no need to do
-	// our own.
-	if tlsConfig.InsecureSkipVerify == false {
-		return tlsConn, nil
+// NewConfigurator creates a new Configurator and sets the provided
+// configuration.
+func NewConfigurator(config Config, logger *log.Logger) (*Configurator, error) {
+	c := &Configurator{logger: logger}
+	err := c.Update(config)
+	if err != nil {
+		return nil, err
 	}
+	return c, nil
+}
 
-	// If verification is not turned on, don't do it.
-	if !c.VerifyOutgoing {
-		return tlsConn, nil
+// Update updates the internal configuration which is used to generate
+// *tls.Config.
+// This function acquires a write lock because it writes the new config.
+func (c *Configurator) Update(config Config) error {
+	cert, err := loadKeyPair(config.CertFile, config.KeyFile)
+	if err != nil {
+		return err
 	}
-
-	if err = tlsConn.Handshake(); err != nil {
-		tlsConn.Close()
-		return nil, err
+	cas, err := loadCAs(config.CAFile, config.CAPath)
+	if err != nil {
+		return err
 	}
 
-	// The following is lightly-modified from the doFullHandshake
-	// method in crypto/tls's handshake_client.go.
-	opts := x509.VerifyOptions{
-		Roots:         tlsConfig.RootCAs,
-		CurrentTime:   time.Now(),
-		DNSName:       "",
-		Intermediates: x509.NewCertPool(),
+	if err = c.check(config, cert); err != nil {
+		return err
 	}
+	c.Lock()
+	c.base = &config
+	c.cert = cert
+	c.cas = cas
+	c.version++
+	c.Unlock()
+	c.log("Update")
+	return nil
+}
 
-	certs := tlsConn.ConnectionState().PeerCertificates
-	for i, cert := range certs {
-		if i == 0 {
-			continue
+func (c *Configurator) check(config Config, cert *tls.Certificate) error {
+	// Check if a minimum TLS version was set
+	if config.TLSMinVersion != "" {
+		if _, ok := TLSLookup[config.TLSMinVersion]; !ok {
+			return fmt.Errorf("TLSMinVersion: value %s not supported, please specify one of [tls10,tls11,tls12]", config.TLSMinVersion)
 		}
-		opts.Intermediates.AddCert(cert)
 	}
 
-	_, err = certs[0].Verify(opts)
-	if err != nil {
-		tlsConn.Close()
-		return nil, err
+	// Ensure we have a CA if VerifyOutgoing is set
+	if config.VerifyOutgoing && config.CAFile == "" && config.CAPath == "" {
+		return fmt.Errorf("VerifyOutgoing set, and no CA certificate provided!")
 	}
 
-	return tlsConn, err
-}
-
-// Configurator holds a Config and is responsible for generating all the
-// *tls.Config necessary for Consul. Except the one in the api package.
-type Configurator struct {
-	sync.Mutex
-	base   *Config
-	checks map[string]bool
+	// Ensure we have a CA and cert if VerifyIncoming is set
+	if config.VerifyIncoming || config.VerifyIncomingRPC || config.VerifyIncomingHTTPS {
+		if config.CAFile == "" && config.CAPath == "" {
```

Comment:
Is it okay to not currently have any CAs if config.CAPath doesn't contain any or when config.CAFile doesn't exist yet?

Is there any reason why you couldn't/shouldn't validate against the loaded CAs?

Reply:
I left it that way because this is the way it is on master, and I tried to make as little changes as possible:
https://github.com/hashicorp/consul/blob/e293045751ecc2b8979bc7bc06fcad7617f0ee29/tlsutil/config.go#L257-L260

If you think it would be worth changing, I would be happy to do that, because your points are absolutely valid.

Reply:
Is it necessary though? I am not sure. Maybe we are relying on validation of the ca config prior to calling this function.

Just bringing it up, if there is other validation in place to ensure that this works correctly then I would add a comment about why it doesn't do more extensive validation here.

Its up to you. If you think this is fine then its probably okay to leave it as is.

Reply:
There is no other validation happening before that. Thinking about this more, I think it is bug:

```
$ cat config.json
{
	"verify_outgoing": true,
	"ca_path": "./empty/"
}
$ consul agent -dev -config-file config.json
==> Starting Consul agent...
==> Consul agent running!
           Version: 'v1.4.3-34-ge29304575-dev (e29304575)'
           Node ID: 'bc56ecf6-ccb3-47c8-c4fb-4b3638545dfe'
         Node name: 'pina'
        Datacenter: 'dc1' (Segment: '<all>')
            Server: true (Bootstrap: false)
       Client Addr: [127.0.0.1] (HTTP: 8500, HTTPS: -1, gRPC: 8502, DNS: 8600)
      Cluster Addr: 127.0.0.1 (LAN: 8301, WAN: 8302)
           Encrypt: Gossip: false, TLS-Outgoing: true, TLS-Incoming: false

==> Log data will now stream in as it occurs:

    2019/03/11 21:17:03 [DEBUG] agent: Using random ID "bc56ecf6-ccb3-47c8-c4fb-4b3638545dfe" as node ID
    2019/03/11 21:17:03 [INFO] raft: Initial configuration (index=1): [{Suffrage:Voter ID:bc56ecf6-ccb3-47c8-c4fb-4b3638545dfe Address:127.0.0.1:8300}]
    2019/03/11 21:17:03 [INFO] raft: Node at 127.0.0.1:8300 [Follower] entering Follower state (Leader: "")
    2019/03/11 21:17:03 [INFO] serf: EventMemberJoin: pina.dc1 127.0.0.1
```

I will change it to actually check. Because what should be happening is something like this: 

```
$ consul agent -dev -config-file config.json -log-level trace
==> Starting Consul agent...
==> Error starting agent: Error loading CA: path "./empty/" has no CAs
```

Reply:
An empty file is detected though:

```
$ consul agent -dev -config-file config.json -log-level trace
==> Starting Consul agent...
==> Error starting agent: Error loading CA File: Couldn't parse PEM in: ca.cert
```

Reply:
Changed it to check if it actually has a CA.